### ISSUE TITLE
Attribution scores for CodeBERT classifier are all zeros

### ISSUE THREAD

**Issue by Jirigesi** [meta-pytorch/captum#978]
Hello, 
Thanks for this awesome work! I am trying to extract the attribution score for a source code-related task. Basically, we use CodeBERT from huggingface to get the code snippet embedding and then classify if the code snippet has defects(0: no defect; 1: defect). 

I tried to follow the https://captum.ai/tutorials/Bert_SQUAD_Interpret to extra attribution score. But what I got is just a bunch of 0s. Can anyone help me with my code?

```python
from transformers import RobertaTokenizer, RobertaConfig, RobertaForSequenceClassification, RobertaModel
import numpy as np
import pandas as pd
import torch
import torch.nn as nn
import torch.nn.functional as F
from captum.attr import visualization as viz
from captum.attr import LayerConductance, LayerIntegratedGradients

code_snippet = '@ _checks_drafts ( draft3 = "ip-address" , draft4 = "ipv4" )  \n def is_ipv4 ( instance ) :  \n if not isinstance ( instance , str_types ) :  \n return True  \n if not _ipv4_re . match ( instance ) :  \n return False  \n return all ( 0 <= int ( component ) <= 255 for component in instance . split ( "." ) )  \n '

device = "cpu"
model_path = "microsoft/codebert-base"
# load model
model = RobertaModel.from_pretrained(model_path)
model.to(device)
model.eval()
model.zero_grad()
# load tokenizer
tokenizer = RobertaTokenizer.from_pretrained(model_path)

ref_token_id = tokenizer.pad_token_id # A token used for generating token reference
sep_token_id = tokenizer.sep_token_id # A token used as a separator between question and text and it is also added to the end of the text.
cls_token_id = tokenizer.cls_token_id # A token used for prepending to the concatenated question-text word sequence

def construct_input_ref_pair(text, ref_token_id, sep_token_id, cls_token_id):
    text_ids = tokenizer.encode(text, add_special_tokens=False)

    # construct input token ids
    input_ids = [cls_token_id] + text_ids + [sep_token_id]

    # construct reference token ids 
    ref_input_ids = [cls_token_id] + [ref_token_id] * len(text_ids) + [sep_token_id]

    return torch.tensor([input_ids], device=device), torch.tensor([ref_input_ids], device=device)

def construct_input_ref_token_type_pair(input_ids, sep_ind=0):
    seq_len = input_ids.size(1)
    token_type_ids = torch.tensor([[0 if i <= sep_ind else 1 for i in range(seq_len)]], device=device)
    ref_token_type_ids = torch.zeros_like(token_type_ids, device=device)# * -1
    return token_type_ids, ref_token_type_ids

def construct_input_ref_pos_id_pair(input_ids):
    seq_length = input_ids.size(1)
    position_ids = torch.arange(seq_length, dtype=torch.long, device=device)
    # we could potentially also use random permutation with `torch.randperm(seq_length, device=device)`
    ref_position_ids = torch.zeros(seq_length, dtype=torch.long, device=device)

    position_ids = position_ids.unsqueeze(0).expand_as(input_ids)
    ref_position_ids = ref_position_ids.unsqueeze(0).expand_as(input_ids)
    return position_ids, ref_position_ids
    
def construct_attention_mask(input_ids):
    return torch.ones_like(input_ids)

def construct_whole_bert_embeddings(input_ids, ref_input_ids, \
                                    token_type_ids=None, ref_token_type_ids=None, \
                                    position_ids=None, ref_position_ids=None):
    input_embeddings = model.bert.embeddings(input_ids, token_type_ids=token_type_ids, position_ids=position_ids)
    ref_input_embeddings = model.bert.embeddings(ref_input_ids, token_type_ids=ref_token_type_ids, position_ids=ref_position_ids)
    
    return input_embeddings, ref_input_embeddings
# Get embeddings 
input_ids, ref_input_ids= construct_input_ref_pair(code_snippet, ref_token_id, sep_token_id, cls_token_id)
token_type_ids, ref_token_type_ids = construct_input_ref_token_type_pair(input_ids)
position_ids, ref_position_ids = construct_input_ref_pos_id_pair(input_ids)
attention_mask = construct_attention_mask(input_ids)

indices = input_ids[0].detach().tolist()
all_tokens = tokenizer.convert_ids_to_tokens(indices)

def predict(inputs, token_type_ids=None, position_ids=None, attention_mask=None):
    out = model(input_ids, attention_mask = attention_mask)[0]
    out = F.softmax(out, dim=1)[0][0][0].unsqueeze(0)
    return out

def custom_forward_func(inputs, token_type_ids=None, position_ids=None, attention_mask=None, position=0):
    out = model(input_ids, attention_mask = attention_mask)[0]
    out = F.softmax(out, dim=1)[0][0][0].unsqueeze(0)
    return out

predict_scores = predict(inputs=input_ids, attention_mask=attention_mask)

lig = LayerIntegratedGradients(custom_forward_func, model.embeddings)

attribution_scores, deltas = lig.attribute(inputs=input_ids,
                                  baselines=ref_input_ids,
                                  additional_forward_args=(token_type_ids, position_ids, attention_mask, 0),
                                  return_convergence_delta=True)


```
Here, all attribution scores are 0s, and the result of attribution_sum is a bunch of 'nan' since it is divided by 0s. 

Looking forward to someone's help! 

Thanks 

Jiri

**Comment by andrewhinh** [comment]
@Jirigesi I'm having the same issue and narrowed it down to the torch.autograd.grad() call in the LayerIntegratedGradients.attribute.gradient_func implementation not computing the gradient between the input and output tensors correctly.

**Comment by craymichael** [comment]
Apologies for the delay.

The forward function in the snippet appears to ignore the `inputs` argument passed by Captum:

```python
def custom_forward_func(inputs, ...):
    out = model(input_ids, attention_mask=attention_mask)[0]
```

Because it always uses the outer `input_ids`, the model output does not depend on Captum's interpolated inputs. If the output is independent of the attributed input, the gradients and attributions will be zero.

You want the forward function to use the argument that Captum passes in. For embedding attribution, one pattern is:

```python
def custom_forward_func(input_ids, attention_mask):
    outputs = classifier_model(input_ids=input_ids, attention_mask=attention_mask)
    return outputs.logits

lig = LayerIntegratedGradients(
    custom_forward_func,
    classifier_model.roberta.embeddings.word_embeddings,
)

attrs = lig.attribute(
    inputs=input_ids,
    baselines=ref_input_ids,
    additional_forward_args=(attention_mask,),
    target=target_class,
)
```

Also make sure you are using the sequence-classification model / classifier head. `RobertaModel` alone returns hidden states, not defect-class logits.

This should answer the question, so I'll close this for now. Please feel free to reply with any follow-up questions.
